packages/api/src/providers/ActivityTyping/ActivityTypingComposer.tsx

```diff
@@ -47,7 +38,12 @@ const ActivityTypingComposer = ({ children }: Props) => {
           type
         } = activity;
 
-        if (type === 'typing' && (role === 'bot' || role === 'user')) {
+        const livestreamingMetadata = getActivityLivestreamingMetadata(activity);
+
+        if (type === 'message' || livestreamingMetadata?.type === 'final activity') {
+          nextTyping.delete(id);
+          changed = true;
+        } else if (type === 'typing' && (role === 'bot' || role === 'user')) {
           const currentTyping = nextTyping.get(id);
```

Comment:
Just throwing some ideas: seems this could also lead to more re-renders due to this change might cause more frequent updates to the typing state

Reply:
Unlikely I guess, here is why.

For `type === 'message' || livestreamingMetadata?.type === 'final activity'`, it means, when receiving a message activity or livestreaming final activity, then, it should remove the typing entry.

However, all final activity *must be* type of message. So, the extraneous `||` here is for defensive programming.

Looking at the whole thing, it would just switched the position of the if-statement to the top rather than the bottom.

Then, the other if-statement is `type === 'typing'`, so it would be orthogonal to `type === 'message'`. So switching the position of the if-statement shouldn't matter.